### PR TITLE
Add support for an unsynthesizable ROB to produce a TracedInstruction stream with wdata

### PR DESCRIPTION
**Pull request by jerryz123**
The ROB is implemented in C++, and interfaced through DPI.
It is unsynthesizable, of course, but it lets us generate a wdata-enabled commit log, if desired.

This is useful for debugging.

**Related issue**: 

**Type of change**: bug report | feature request | other enhancement

**Impact**: no functional change | API addition (no impact on existing code) | API modification

**Development Phase**: proposal |  implementation

**Release Notes**

### CODE REVIEW

File: src/main/scala/rocket/RocketCore.scala [hunk 1]
```diff
@@ -53,7 +53,8 @@ case class RocketCoreParams(
   mvendorid: Int = 0, // 0 means non-commercial implementation
   mimpid: Int = 0x20181004, // release date in BCD
   mulDiv: Option[MulDivParams] = Some(MulDivParams()),
-  fpu: Option[FPUParams] = Some(FPUParams())
+  fpu: Option[FPUParams] = Some(FPUParams()),
+  debugROB: Boolean = false // if enabled, uses a C++ debug ROB to generate trace-with-wdata
```
Comment (sequencer) [inline src/main/scala/rocket/RocketCore.scala:57]: This is fine before Chisel5, I think eventually we need split the test(unsynthesisable) codes out from RTL with the Probe API.

**Review by sequencer — COMMENTED** [review]
This PR looks good, cc @ZenithalHourlyRate for detail review.
There is another methodology for this implementation: probe the scoreboard to record the order and reorder by it. Just propose this method for discussion.

File: src/main/scala/rocket/RocketCore.scala (hunk 1 above)
Comment (jerryz123) [reply src/main/scala/rocket/RocketCore.scala]: We can just replace DebugROB backend with XMR instead of DPI when that time comes

**Comment by jerryz123** [comment]
@sequencer are there any logs from CI?

**Comment by jerryz123** [comment]
@ZenithalHourlyRate thoughts?

**Review by ZenithalHourlyRate — APPROVED** [review]
Have read through the code and LGTM.

https://github.com/chipsalliance/rocket-chip/blob/e65e1132ed2f842cc0168bd2d98e712db4f35b8f/src/main/scala/rocket/RocketCore.scala#L1017-L1042

I think we could add similar code to this when with both DebugROB and enable commit log? these `0xXXXXXXXXXXXXXXXX` and `when (ll_wen && rf_waddr =/= 0.U)` are not needed now. Or this should be handled by `TraceGenConfig`

File: src/main/scala/rocket/DebugROB.scala [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    val valid = Input(Bool())
+    val wb_tag = Input(UInt(64.W))
+    val wb_data = Input(UInt(512.W))
+  })
+  addResource("/csrc/debug_rob.cc")
+  addResource("/vsrc/debug_rob.v")
+}
+
+class DebugROBPopTrace(implicit val p: Parameters) extends BlackBox with HasBlackBoxResource with HasCoreParameters {
+  require(traceHasWdata && (vLen max xLen) <= 512)
```
[the remark this answers is not in the document]
Comment (jerryz123) [reply src/main/scala/rocket/DebugROB.scala:60]: Given that there is no available vector unit to take advantage of the vector wdata of trace, this shouldn't matter for now.

[the remark this answers is not in the document]
Comment (ZenithalHourlyRate) [reply src/main/scala/rocket/DebugROB.scala:60]: Agreed.

**Pull request merged**